### PR TITLE
fix(log): Attribute third-party publish to organization with import source #NP-50757

### PR DESCRIPTION
**Pull request by torbjokv**
When a publication was published by a third-party/backend client-credentials client, the log entry recorded `performedBy.type = "Person"` with the client's acting-user name, instead of attributing it to the organization with an import source. Unlike `CreatedResourceEvent` and `UpdatedResourceEvent`, `PublishedResourceEvent` never captured the client's import source, so `LogEntryService` always fell back to building a `LogUser` ("Person").

- Add an `importSource` field to `PublishedResourceEvent`, populated from the client's `ThirdPartySystem` when the user is an external client (symmetric with create/update events)
- Add a `PublishedResourceEvent` branch in `LogEntryService` so a published event carrying an import source is attributed to a `LogOrganization` ("Organization")
- Add regression test verifying a third-party publish is logged as Organization with a non-null import source

https://sikt.atlassian.net/browse/NP-50757

### CODE REVIEW

File: publication-commons/src/main/java/no/unit/nva/publication/model/business/UserInstance.java [hunk 1]
```diff
@@ -214,6 +216,16 @@ public Optional<ThirdPartySystem> getThirdPartySystem() {
     return isExternalClient() ? Optional.ofNullable(thirdPartySystem) : Optional.empty();
   }
 
+  public ImportSource getImportSource() {
```
Comment (brinxmat) [inline publication-commons/src/main/java/no/unit/nva/publication/model/business/UserInstance.java]: Should this be `Optional<ImportSource>` ?

File: publication-commons/src/test/java/no/unit/nva/publication/model/business/publicationstate/ResourceEventTest.java [hunk 2]
```diff
@@ -46,7 +46,11 @@ public static Stream<Arguments> resourceEventProvider() {
                 ImportSource.fromSource(Source.OTHER)),
             PUBLICATION_CREATED),
         Arguments.of(
-            new PublishedResourceEvent(Instant.now(), randomUser(), randomUri(), identifier()),
+            new PublishedResourceEvent(
+                Instant.now(), randomUser(), randomUri(), identifier(), null),
```
Comment (brinxmat) [inline publication-commons/src/test/java/no/unit/nva/publication/model/business/publicationstate/ResourceEventTest.java]: This null could be an explicitly named constant.

File: publication-commons/src/test/java/no/unit/nva/publication/model/business/ResourceEventTest.java [hunk 3]
```diff
@@ -41,7 +41,11 @@ public static Stream<Arguments> stateProvider() {
                 Instant.now(), new User(randomString()), randomUri(), SortableIdentifier.next())),
         Arguments.of(
             new PublishedResourceEvent(
-                Instant.now(), new User(randomString()), randomUri(), SortableIdentifier.next())),
+                Instant.now(),
+                new User(randomString()),
+                randomUri(),
+                SortableIdentifier.next(),
+                null)),
```
Comment (brinxmat) [inline publication-commons/src/test/java/no/unit/nva/publication/model/business/ResourceEventTest.java]: Again a named constant makes things clearer

File: publication-log/src/main/java/no/unit/nva/publication/log/service/LogEntryService.java [hunk 4]
```diff
@@ -101,7 +102,9 @@ private LogEntry createLogEntry(Resource resource, ResourceEvent resourceEvent)
         || resourceEvent instanceof MergedResourceEvent
         || resourceEvent instanceof CreatedResourceEvent event && nonNull(event.importSource())
         || resourceEvent instanceof UpdatedResourceEvent updatedResourceEvent
-            && nonNull(updatedResourceEvent.importSource())) {
+            && nonNull(updatedResourceEvent.importSource())
+        || resourceEvent instanceof PublishedResourceEvent publishedResourceEvent
+            && nonNull(publishedResourceEvent.importSource())) {
```
Comment (brinxmat) [inline publication-log/src/main/java/no/unit/nva/publication/log/service/LogEntryService.java]: This looks like a code smell. 

We should refactor to use composition, not this kind of laundry list.

Maybe refactor straight away.

File: publication-log/src/test/java/no/unit/nva/publication/log/service/LogEntryServiceTest.java [hunk 5]
```diff
@@ -202,6 +206,27 @@ void shouldPersistFileUploadedLogEntryForThirdParty() throws BadRequestException
...
+
+    logEntryService.persistLogEntry(Resource.fromPublication(publication));
+
+    var logEntry =
+        (PublicationLogEntry)
+            Resource.fromPublication(publication).fetchLogEntries(resourceService).getFirst();
+
+    assertEquals(PUBLICATION_PUBLISHED, logEntry.topic());
+    assertNotNull(logEntry.importSource());
+    assertInstanceOf(LogOrganization.class, logEntry.performedBy());
```
Comment (brinxmat) [inline publication-log/src/test/java/no/unit/nva/publication/log/service/LogEntryServiceTest.java]: Are there any expectations about the LogOrganization?

Perhaps compare apples and apples in this case?

`var expected = new LogOrganization...`

File: publication-log/src/test/java/no/unit/nva/publication/log/service/LogEntryServiceTest.java [hunk 6]
```diff
@@ -202,6 +206,27 @@ void shouldPersistFileUploadedLogEntryForThirdParty() throws BadRequestException
     assertNotNull(logEntry.importSource());
   }
 
+  @Test
+  void shouldPersistPublishedLogEntryAsOrganizationForThirdParty() throws BadRequestException {
```
Comment (brinxmat) [inline publication-log/src/test/java/no/unit/nva/publication/log/service/LogEntryServiceTest.java]: Perhaps `shouldPersistLogEntryAsOrganizationForThirdPartyImports`

**Review by brinxmat — DISMISSED** [review]

File: publication-log/src/test/java/no/unit/nva/publication/log/service/LogEntryServiceTest.java (hunk 6 above)
Comment (torbjokv) [reply publication-log/src/test/java/no/unit/nva/publication/log/service/LogEntryServiceTest.java]: shouldPersistLogEntryAsOrganizationForThirdPartyPublish!

File: publication-log/src/main/java/no/unit/nva/publication/log/service/LogEntryService.java
Comment (LarsV123) [inline publication-log/src/main/java/no/unit/nva/publication/log/service/LogEntryService.java:153]: Unrelated to this PR, but I noticed it now... Silently swallowing all exceptions seems like it could cause confusion at some point. We should at least log it.

Comment (LarsV123) [inline publication-log/src/main/java/no/unit/nva/publication/log/service/LogEntryService.java:152]:
```suggestion
      logger.warn("Failed to enrich LogUser for user {} (institution {}); falling back to minimal record", user, institution, e);
      return LogUser.fromResourceEvent(user, institution);
```

File: publication-log/src/test/java/no/unit/nva/publication/log/service/LogEntryServiceTest.java [hunk 7]
```diff
@@ -202,6 +205,32 @@ void shouldPersistFileUploadedLogEntryForThirdParty() throws BadRequestException
     assertNotNull(logEntry.importSource());
   }
 
+  @Test
+  void shouldPersistLogEntryAsOrganizationForThirdPartyPublish() throws BadRequestException {
```
Comment (LarsV123) [inline publication-log/src/test/java/no/unit/nva/publication/log/service/LogEntryServiceTest.java:210]: Maybe add a negative test to assert a regular user publish is properly attributed to `LogUser` and not a `LogOrganization`?

File: publication-commons/src/main/java/no/unit/nva/publication/model/business/publicationstate/FileUploadedEvent.java
Comment (LarsV123) [inline publication-commons/src/main/java/no/unit/nva/publication/model/business/publicationstate/FileUploadedEvent.java]: Should this also implement `ImportSourceProvider`?

**Review by LarsV123 — APPROVED** [review]

File: publication-commons/src/main/java/no/unit/nva/publication/model/business/publicationstate/FileUploadedEvent.java
Comment (torbjokv) [reply publication-commons/src/main/java/no/unit/nva/publication/model/business/publicationstate/FileUploadedEvent.java]: good point!
